modulus/utils/zenith_angle.py

```diff
@@ -101,6 +105,174 @@ def cos_zenith_angle_from_timestamp(
     return _star_cos_zenith(julian_centuries, lon_rad, lat_rad)
 
 
+def irradiance(
+    t, S0=1361, e=0.0167, perihelion_longitude=282.895, mean_tropical_year=365.2422
+):
+    """The flux of solar energy in W/m2 towards Earth
+
+    The default orbital parameters are set to 2000 values.
+    Over the period of 1900-2100 this will result in an error of at most 0.02%,
+    so can be neglected for many applications.
+
+    Args:
+        t: linux timestamp
+        S0: the solar constant in W/m2. This is the mean irradiance received by
+            earth over a year.
+        e: the eccentricity of earths elliptical orbit
+        perihelion_longitude: spatial angle from moving vernal equinox to perihelion with Sun as angle vertex.
+            Perihelion is moment when earth is closest to sun. vernal equinox is
+            the longitude when the Earth crosses the equator from South to North.
+
+
+    Notes:
+
+        TISR can be computed from Berger's formulas:
+
+            Berger, A. (1978). Long-Term Variations of Daily Insolation and
+            Quaternary Climatic Changes. Journal of the Atmospheric Sciences,
+            35(12), 2362–2367.
+            https://doi.org/10.1175/1520-0469(1978)035<2362:LTVODI>2.0.CO;2
+
+        NASA Example computing the orbital parameters: https://data.giss.nasa.gov/modelE/ar5plots/srorbpar.html. From 1900-2100 these are the ranges::
+            Orbital Parmameters
+
+                                                Long. of
+            Year     Eccentri    Obliquity    Perihel.
+            (A.D.)      city      (degrees)    (degrees)
+            ------    --------    ---------    --------
+                1900   0.016744      23.4528      281.183
+                2000   0.016704      23.4398      282.895
+                2100   0.016663      23.4268      284.609
+
+    """
+    seconds_per_solar_day = 86400
+    mean_tropical_year = mean_tropical_year * seconds_per_solar_day
+
+    year_2000_equinox = datetime.datetime(2000, 3, 20, 7, 35, tzinfo=pytz.utc)
+
+    # from appendix of Berger 1978
+    lam = (t - year_2000_equinox.timestamp()) % mean_tropical_year
+    lam = lam / mean_tropical_year * 2 * np.pi
+    v = lam - perihelion_longitude * np.pi / 180
```

Comment:
This is the mean anomaly, isn't it? But the next line computes `rho` (which ideally should be renamed to `rho_inv`) but the formula should have the true anomaly which is very slightly different from the mean anomaly for small eccentricities like it is for Earth. See section 6.8 here: https://link.springer.com/book/10.1007/978-3-662-53045-0

Standard way is to go from the mean anomaly to the eccentric anomaly and then to the true anomaly, but this doesn't have a closed form solution so would need a nonlinear solver.

Reply:
Thanks. It would be nice to get the plot to line up exactly. Sec 3 of [Berger](https://journals.ametsoc.org/view/journals/atsc/35/12/1520-0469_1978_035_2362_ltvodi_2_0_co_2.xml?tab_body=pdf) also describes some iterative procedure, but I skimmed over that part.